c/ckb_mmr.h

```diff
@@ -0,0 +1,474 @@
+                                   const uint8_t *buffer,
+                                   uint32_t buffer_length) {
+  context->buffer = buffer;
+  context->buffer_length = buffer_length;
+  context->index = 0;
+  return 0;
+}
+
+int _mmr_default_buffer_read_node(void *context, mmr_node_t *out) {
+  mmr_default_buffer_reader_t *c = (mmr_default_buffer_reader_t *)context;
+
+  if (c->buffer_length - c->index < 2) {
+    return ERROR_NODE_EOF;
+  }
+  uint16_t len = *((uint16_t *)(&c->buffer[c->index]));
+  if (c->buffer_length - c->index - ((uint32_t)2) < len) {
+    return ERROR_NODE_EOF;
+  }
+  out->value.pointer = &c->buffer[c->index + 2];
+  out->t = MMR_NODE_TYPE_POINTER;
+  out->length = len;
+  c->index += 2 + len;
+  return 0;
+}
+
+inline int _mmr_default_command_read(void *context, uint8_t *out_command) {
+  mmr_default_buffer_reader_t *c = (mmr_default_buffer_reader_t *)context;
+
+  if (c->index >= c->buffer_length) {
+    return ERROR_NO_MORE_COMMANDS;
+  }
+  *out_command = c->buffer[c->index++];
+  return 0;
+}
+
+int _mmr_default_leaf_iter_read(void *context, mmr_node_t *out_node,
+                                uint64_t *out_position) {
+  mmr_default_buffer_reader_t *c = (mmr_default_buffer_reader_t *)context;
+
+  if (c->index >= c->buffer_length) {
+    return ERROR_NO_MORE_LEAFS;
+  }
+  if (c->buffer_length - c->index < 8) {
+    return ERROR_LEAF_EOF;
+  }
+  uint64_t position = *((uint64_t *)(&c->buffer[c->index]));
+  c->index += 8;
+
+  int ret = _mmr_default_buffer_read_node(c, out_node);
+  if (ret != 0) {
+    return ret;
+  }
+  *out_position = position;
+  return 0;
+}
+
+#ifdef __CKB_FORCE_RISCV_B__
+uint64_t _mmr_trailing_zeros_u64(uint64_t value) {
+  uint64_t ret;
+  asm("ctz %0,%1" : "=r"(ret) : "r"(value));
+  return ret;
+}
+
+uint64_t _mmr_leading_zeros_u64(uint64_t value) {
+  uint64_t ret;
+  asm("clz %0,%1" : "=r"(ret) : "r"(value));
+  return ret;
+}
+
+uint64_t _mmr_count_zeros_u64(uint64_t value) {
+  uint64_t ret;
+  asm("cpop %0,%1" : "=r"(ret) : "r"(value));
+  return 64 - ret;
+}
+#elif defined __has_builtin
+uint64_t _mmr_trailing_zeros_u64(uint64_t value) {
+  return __builtin_ctzl(value);
+}
+
+uint64_t _mmr_leading_zeros_u64(uint64_t value) {
+  return __builtin_clzl(value);
+}
+
+uint64_t _mmr_count_zeros_u64(uint64_t value) {
+  return 64 - __builtin_popcountl(value);
+}
+#else
+#error "No implementation available for ctz, clz and cpop!"
+#endif
+
+uint64_t _mmr_parent_offset(uint32_t height) { return 2 << height; }
+
+uint64_t _mmr_sibling_offset(uint32_t height) { return (2 << height) - 1; }
+
+int _mmr_all_ones(uint64_t num) {
+  return num != 0 && _mmr_count_zeros_u64(num) == _mmr_leading_zeros_u64(num);
+}
+
+uint64_t _mmr_jump_left(uint64_t pos) {
+  uint64_t bit_length = 64 - _mmr_leading_zeros_u64(pos);
+  uint64_t most_significant_bits = ((uint64_t)1) << (bit_length - 1);
+  return pos - (most_significant_bits - 1);
+}
+
+uint32_t _mmr_pos_height_in_tree(uint64_t pos) {
```

Comment:
The arg `pos` is extracted from user input proof. It causes overflow in `_mmr_jump_left` when pos equal to `(uint64_t)(0 - 1)`.